[jit] make profiling take no_grad flags into account

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31071 [jit] make profiling take no_grad flags into account**

Previously the profiler would think Tensors would require grad, even
when the no_grad flag is enabled during execution. This makes the profiling
and guards respect the no_grad flag, which eliminates extra differentiable
graphs that appear in the backward graph (where no_grad is typically enabled).

Differential Revision: [D18915468](https://our.internmc.facebook.com/intern/diff/D18915468)

zdevito committed:
[jit] make profiling take no_grad flags into account

Previously the profiler would think Tensors would require grad, even
when the no_grad flag is enabled during execution. This makes the profiling
and guards respect the no_grad flag, which eliminates extra differentiable
graphs that appear in the backward graph (where no_grad is typically enabled).

kostmo commented:
## CircleCI build failures summary
As of commit 68774d9:
* **1/1** broken upstream at merge base 06c7420 (see [grid view](https://dr.pytorch.org/master-timeline.html?count=10&sha1=06c7420fa24235b2bdbf4e4abe7ca20d291f86aa&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * If your commit is older than `viable/strict`:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
        * If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
          ```
          git fetch viable/strict
          git rebase --onto viable/strict $(git merge-base origin/master HEAD)
          ```
* **0/1** failures introduced in this PR



## Detailed failure analysis (WIP)

Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=68774d91365982b31f9bcb10c23338a3e1e41824):

### 1 failure recognized by patterns:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3886618) pytorch_linux_xenial_py2_7_9_build (1/1)
**Step:** Build ([details](https://dr.pytorch.org/build-details.html?build_id=71845967))
```
Dec 10 20:45:57 /var/lib/jenkins/workspace/third_party/fbgemm/include/fbgemm/FbgemmFP16.h:243:3: error: type qualifiers ignored on function return type [-Werror=ignored-qualifiers]
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

Krovatkin reviewed (approved):
:shipit:

zdevito committed:
Update on "[jit] make profiling take no_grad flags into account"

[jit] make profiling take no_grad flags into account

Previously the profiler would think Tensors would require grad, even
when the no_grad flag is enabled during execution. This makes the profiling
and guards respect the no_grad flag, which eliminates extra differentiable
graphs that appear in the backward graph (where no_grad is typically enabled).

gh-metadata: pytorch pytorch 31071 gh/zdevito/139/head
